src/Preprocessor.zig

```diff
@@ -408,7 +408,16 @@ fn expr(pp: *Preprocessor, tokenizer: *Tokenizer) Error!bool {
             } else {
                 tok.id = .zero;
             }
+        } else if (pp.tokens.len == start and !tok.id.validPreprocessorExprStart()) {
+            try pp.err(tok, .invalid_preproc_expr_start);
+            skipToNl(tokenizer);
+            return false;
+        } else if (tok.id.invalidPreprocessorBinOp()) {
+            try pp.err(tok, .invalid_preproc_operator);
+            skipToNl(tokenizer);
+            return false;
```

Comment:
This won't work you have to put the check in the switch below.
```c
#define FOO --
#if 0 FOO
#endif
```

Reply:
Made the requested change + rebased against master. One thing I noticed is that diagnostics generated from within that switch can end up pointing to the wrong token:
```c
int foo;
#if "test"
#endif
```
```
test.c:1:1: error: string literal in preprocessor expression
int foo;
^
```

Reply:
I think maybe we just need `pp.tokens.items(.loc)[start + i]` instead of `pp.tokens.items(.loc)[i]` for error locations